### PR TITLE
Merge StoragePackerV2 implementation

### PR DESCRIPTION
**Pull request by mgritter**
Original PR at https://github.com/hashicorp/vault/pull/6874

### CODE REVIEW

File: sdk/helper/storagepacker/ids2keys.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/vault/sdk/helper/cryptoutil"
+	"math"
+	"sort"
+	"strings"
+)
+
+type itemRequest struct {
```
Comment (calvn) [inline sdk/helper/storagepacker/ids2keys.go]: Why is this struct needed as opposed to simply using `Item` directly? The only additional info in here is the hash which is can be calculated using the ID.

File: sdk/helper/storagepacker/invalidate.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+		bucketRaw, found := s.bucketsCache.Get(k)
+		if found {
+			parents = append(parents, bucketRaw.(*LockedBucket))
+		}
+	}
+	return parents
+}
+
+// Given a bucket key, return the cache key for its parent
+func (s *StoragePackerV2) parentCacheKey(bucketKey string) string {
```
Comment (calvn) [inline sdk/helper/storagepacker/invalidate.go]: What happens if this is called against the topmost bucket?

File: sdk/helper/storagepacker/ids2keys.go (hunk 1 above)
Comment (mgritter) [reply sdk/helper/storagepacker/ids2keys.go]: Recalculating the hash every time we needed it seemed like a bad idea. For example, sorting in Key order would become more expensive.

But, yeah, I originally thought it was going to have a pointer to the Bucket too but that turned out to be unnecessary.

File: sdk/helper/storagepacker/invalidate.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+	if originalBucket.Key == replacementBucket.Key {
+		// Any item not present in the new bucket *might* be deleted, or
+		// it might be elsewhere (in which case we want to report that value instead,
+		// to make sure the correct version is being used.)
+		maybeDeleted := itemDifferenceBetween(originalBucket, replacementBucket)
+		var revisit []*Item
+		deleted, revisit, err = s.identifyItemsAbsentOrShadowed(bucketKey, maybeDeleted)
+		if err != nil {
+			// Return partial information?
+			return present, nil, err
```
[the remark this answers is not in the document]
Comment (mgritter) [reply sdk/helper/storagepacker/invalidate.go]: I don't have a good for what the invalidation code ought to do when there is a partial error.

Does it just shut down the process and try to resync? Attempt to keep going? If the latter, then the partial result seems useful even if we don't end up deleting anything.

Returning nil isn't a big deal, but I thought neither was returning the slice we've already created.

File: sdk/helper/storagepacker/ids2keys.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func GetItemIDHash(itemID string) string {
+	return hex.EncodeToString(cryptoutil.Blake2b256Hash(itemID))
+}
+
+// Length of key
+const KeyLength = 64
+
+// Given a list of IDs, calculate their keys generate itemRequests for each.
```
[the remark this answers is not in the document]
Comment (mgritter) [reply sdk/helper/storagepacker/ids2keys.go]: OK, I wasn't aware this was the go convention.

File: sdk/helper/storagepacker/invalidate.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+}
+
+// Look up each item and determine whether another copy (or version) of it exists
+// in a different bucket.
+func (s *StoragePackerV2) identifyItemsAbsentOrShadowed(bucketKey string, maybeDeleted []*itemRequest) ([]*Item, []*Item, error) {
+	deleted := make([]*Item, 0)
+	revisit := make([]*Item, 0)
+
+	// Using the radix lookup will find child buckets, but not parent
```
[the remark this answers is not in the document]
Comment (mgritter) [reply sdk/helper/storagepacker/invalidate.go]: Good suggestion, I missed that.

File: sdk/helper/storagepacker/invalidate.go (hunk 2 above)
Comment (mgritter) [reply sdk/helper/storagepacker/invalidate.go]: It returns "".

File: sdk/helper/storagepacker/storagepacker_v2.go [hunk 6]
```diff
@@ -0,0 +1,716 @@
...
+	})
+	s.bucketsCacheLock.RUnlock()
+
+	return ret, nil
+}
+
+// Parse a bucket from storage, but don't add it to the cache here.
+// (We may want to read a v1 bucket and upgrade it?)
+func (s *StoragePackerV2) DecodeBucket(storageEntry *logical.StorageEntry) (*LockedBucket, error) {
+	uncompressedData, notCompressed, err := compressutil.Decompress(storageEntry.Value)
```
[the remark this answers is not in the document]
Comment (mgritter) [reply sdk/helper/storagepacker/storagepacker_v2.go]: OK, will fix.

File: sdk/helper/storagepacker/partition.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+// Requests for multi-item put or get, partitioned by bucket.
+// This lets us operate bucket-by-bucket.
+type partitionedRequests struct {
+	Bucket   *LockedBucket
+	Requests []*itemRequest
+}
+
+// Take the list of requests and partition them by which bucket the
+// Key belongs to.
```
[the remark this answers is not in the document]
Comment (mgritter) [reply sdk/helper/storagepacker/partition.go]: I really meant Key (the field) here so I think capitalization is better.

File: sdk/helper/storagepacker/partition.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+func (s *StoragePackerV2) partitionRequests(unsortedRequests []*itemRequest) ([]*partitionedRequests, error) {
+	sortedRequests := sortRequests(unsortedRequests)
+
+	partition := make([]*partitionedRequests, 0)
+	var lastBucket *partitionedRequests
+
+	// The items requests are sorted by key, so the same buckets end
+	// up together--- but the radix tree doesn't have a way to take
```
[the remark this answers is not in the document]
Comment (mgritter) [reply sdk/helper/storagepacker/partition.go]: What have you got against my em-dashes?  :)

I'll use your preferred punctuation.

File: sdk/helper/storagepacker/invalidate.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgritter) [reply sdk/helper/storagepacker/invalidate.go]: Done.

File: sdk/helper/storagepacker/storagepacker_v2.go [hunk 9]
```diff
@@ -0,0 +1,726 @@
...
+	*locksutil.LockEntry
+}
+
+func (s *StoragePackerV2) preloadBucketsFromDisk(ctx context.Context, keys []string) error {
+	// If we crash while writing out new buckets created by sharding,
+	// we will have version N of the bucket in storage, and version N+1
+	// represented in the sub-buckets.
+	//
+	// To bring these two into alignment, we can roll back to version N.
+	// It might be possible to foll forward in some cases (particularly if
```
Comment (calvn) [inline sdk/helper/storagepacker/storagepacker_v2.go]:
```suggestion
	// It might be possible to roll forward in some cases (particularly if
```

File: sdk/helper/storagepacker/ids_test.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"context"
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/sdk/logical"
+	"testing"
+)
+
+func getStoragePackerBits(tb testing.TB, baseBucketBits int, bucketShardBits int) *StoragePackerV2 {
+	storage := &logical.InmemStorage{}
+	storageView := logical.NewStorageView(storage, "packer/buckets/v2")
```
Comment (calvn) [inline sdk/helper/storagepacker/ids_test.go]: I think this should be `logical.NewStorageView(storage, StoragePackerBucketsPrefix)`. `NewStoragePackerV2` operates on `config.BucketStorageView.Subview("v2/")` ([link](https://github.com/hashicorp/vault/pull/6724/files#diff-d460b49b4e599a77c84c96a74553b2d8R674)) so it's currently writing to `packer/buckets/v2/v2/`.

File: sdk/helper/storagepacker/storagepacker_v2.go [hunk 11]
```diff
@@ -0,0 +1,726 @@
...
+		return nil, retErr
+	}
+
+	ret := make([]string, 0, 256)
+	s.bucketsCacheLock.RLock()
+	s.bucketsCache.Walk(func(s string, _ interface{}) bool {
+		ret = append(ret, s)
+		return false
+	})
+	s.bucketsCacheLock.RUnlock()
```
Comment (calvn) [inline sdk/helper/storagepacker/storagepacker_v2.go]: I think this `RUnlock` should be inside Do's `func`.

File: sdk/helper/storagepacker/invalidate.go [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+// MatchingStorage returns true if the given storage path is
+// controlled by this StoragePacker.
+func (s *StoragePackerV2) MatchingStorage(path string) bool {
+	return strings.HasPrefix(path, s.BucketStorageView.Prefix())
+}
+
+func (s *StoragePackerV2) bucketKeyFromPath(path string) (string, error) {
+	bucketRoot := s.BucketStorageView.Prefix()
+
+	if strings.HasPrefix(path, bucketRoot) {
```
Comment (calvn) [inline sdk/helper/storagepacker/invalidate.go]: You can simply use `if s.MatchingStorage() {...}` in here.

File: sdk/helper/storagepacker/invalidate.go [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+		bucketRemoved = true
+		// Create a placeholder so we can run the rest of the logic
+		// against an empty collection
+		replacementBucket = s.newEmptyBucket(bucketKey)
+	} else {
+		// Fake up a storage entry for the bucket and decode it
+		storage := &logical.StorageEntry{
+			Value: newValue,
+		}
+		replacementBucket, err = s.DecodeBucket(storage)
```
Comment (calvn) [inline sdk/helper/storagepacker/invalidate.go]: Wouldn't the cacheKey in the Decode call result in an empty `cacheKey`? I think it'd lead to weird locking behavior due to this value being empty in `locksutil.LockForKey(s.storageLocks, cacheKey)`.

File: sdk/helper/storagepacker/storagepacker_v2.go [hunk 14]
```diff
@@ -0,0 +1,726 @@
...
+
+	return ret, nil
+}
+
+// DecodeBucket parses a bucket from storage, but doesn't add it to the cache.
+// (We may want to read a v1 bucket and upgrade it?)
+func (s *StoragePackerV2) DecodeBucket(storageEntry *logical.StorageEntry) (*LockedBucket, error) {
+	if storageEntry == nil || storageEntry.Value == nil {
+		return nil, errors.New("decoding nil storageEntry")
```
Comment (calvn) [inline sdk/helper/storagepacker/storagepacker_v2.go]:
```suggestion
		return nil, errors.New("decoding nil storage entry")
```

File: sdk/helper/storagepacker/ids2keys.go [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+	// Value is the stored object, nil if not found
+	Value *Item
+}
+
+func GetItemIDHash(itemID string) string {
+	return hex.EncodeToString(cryptoutil.Blake2b256Hash(itemID))
+}
+
+// Length of key
+const KeyLength = 64
```
Comment (calvn) [inline sdk/helper/storagepacker/ids2keys.go]: Seems that we don't have to export this. Probably worth mentioning that this is the length returned by blake2b's hashing algo.

File: sdk/helper/storagepacker/storagepacker_v2.go (hunk 11 above)
Comment (mgritter) [reply sdk/helper/storagepacker/storagepacker_v2.go]: Why?

preloadBucketsFromDisk does its own (write-)locking on bucketsCacheLock, and bucketsCacheLock is not acquired in the anonymous function itself.

Comment (calvn) [reply sdk/helper/storagepacker/storagepacker_v2.go]: > bucketsCacheLock is not acquired in the anonymous function itself.

You are calling `s.bucketsCacheLock.RLock()` inside the lambda func, in [L175](https://github.com/hashicorp/vault/pull/6998/files#diff-d460b49b4e599a77c84c96a74553b2d8R175).

Comment (calvn) [reply sdk/helper/storagepacker/storagepacker_v2.go]: Oh wait nvm...

File: sdk/helper/storagepacker/ids_test.go (hunk 10 above)
Comment (mgritter) [reply sdk/helper/storagepacker/ids_test.go]: You're correct, though it doesn't matter here (the redundant v2 isn't used, and would work fine even if we did write data.)

File: sdk/helper/storagepacker/invalidate.go (hunk 13 above)
Comment (mgritter) [reply sdk/helper/storagepacker/invalidate.go]: I believe you're correct, good catch.

DecodeBucket should probably use the demarshaled bucket key rather than the storage key.

Comment (mgritter) [reply sdk/helper/storagepacker/invalidate.go]: OK, fixed 2x (and made sure both pass the new test.)

1. DecodeBucket uses the Bucket's key if it's present
2. InvalidateItems provides a valid Key in the StorageEntry

File: sdk/helper/storagepacker/ids_test.go (hunk 10 above)
Comment (mgritter) [reply sdk/helper/storagepacker/ids_test.go]: fixed

File: sdk/helper/storagepacker/storagepacker_v2.go (hunk 14 above)
Comment (mgritter) [reply sdk/helper/storagepacker/storagepacker_v2.go]: I prefer the actual name of the argument here.

File: sdk/helper/storagepacker/ids2keys.go (hunk 15 above)
Comment (mgritter) [reply sdk/helper/storagepacker/ids2keys.go]: You're correct that it's not strictly necessary. Let me think about it, it's not really the case that we need to export GetItemIdHash either. But, we should probably export both or neither.

The key is a little bit of a leaky abstraction; we still expose buckets (via GetBucket) and if you want to interact with a bucket you need to be able to create a key.

File: sdk/helper/storagepacker/partition.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+// bucket the Key belongs to.
+func (s *StoragePackerV2) partitionRequests(unsortedRequests []*itemRequest) ([]*partitionedRequests, error) {
+	sortedRequests := sortRequests(unsortedRequests)
+
+	partition := make([]*partitionedRequests, 0)
+	var lastBucket *partitionedRequests
+
+	// The items requests are sorted by key, so the same buckets end
+	// up together. But, the radix tree doesn't have a way to take
+	// advantage of that to do a single pass.
```
Comment (jefferai) [inline sdk/helper/storagepacker/partition.go]: Doesn't this mean then that we shouldn't sort it? We're not benefiting but we're taking the efficiency hit.

File: sdk/helper/storagepacker/ids2keys.go [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+// keysForIDs calculates keys and generates itemRequests for each input ID.
+func (s *StoragePackerV2) keysForIDs(ids []string) []*itemRequest {
```
Comment (jefferai) [inline sdk/helper/storagepacker/ids2keys.go]: Why use keys everywhere when the functions deal with requests? e.g. why not `requestsForIDs` here and `requestsForItems` below and naming the file `ids2requests.go`?

Also given that not all functions in here are dealing with converting ids to requests, perhaps a better name would be something like `util.go`.

File: sdk/helper/storagepacker/types.proto [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+// abstraction contains multiple buckets of the same kind within itself and
+// shares amont them the items that get inserted. When the bucket as a whole
+// gets too big to hold more items, the contained buckets gets pushed out only
+// to become independent buckets. Hence, this can grow infinitely in terms of
+// storage space for items that get inserted.
+message Bucket {
+	// Key is the storage path where the bucket gets stored
+	string key = 1;
+	// Items holds the items contained within this bucket. Used by v1.
+	repeated Item items = 2;
```
Comment (jefferai) [inline sdk/helper/storagepacker/types.proto]: Why is this here? This is a different lib altogether so it doesn't seem like we need this field.

File: sdk/helper/storagepacker/types.proto [hunk 19]
```diff
@@ -0,0 +1,40 @@
+syntax = "proto3";
+
+option go_package = "github.com/hashicorp/vault/sdk/helper/storagepacker";
+
+package storagepacker;
+
+import "google/protobuf/any.proto";
+
+// Item represents an entry that gets inserted into the storage packer
+message Item {
```
Comment (jefferai) [inline sdk/helper/storagepacker/types.proto]: Given that v2 is using raw bytes instead of this Item message it seems like we don't need this type and all of the related functions at all?

File: sdk/helper/storagepacker/ids2keys.go [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/vault/sdk/helper/cryptoutil"
+	"math"
+	"sort"
+	"strings"
+)
+
+type itemRequest struct {
```
Comment (jefferai) [inline sdk/helper/storagepacker/ids2keys.go]: Why call it itemRequest? What does the request refer to here?

File: sdk/helper/storagepacker/partition.go (hunk 16 above)
Comment (mgritter) [reply sdk/helper/storagepacker/partition.go]: While sorting doesn't change the number of radix tree lookups we have to do, it does mean that we can produce the buckets in-order, and I do take advantage of that.

I think now it is possible to make the partitioning somewhat more efficient but it would be a little more complex.  Because we always fully shard a bucket, once we do a lookup, we can just compare whether the same-length prefix of the next key is the same.  If the radix tree were less regular that wouldn't work.

So, for example, if our keys are 
AABB
AABC
AACD
ABFF
then if radix lookup returns AA/B we can take the first two keys and do our next lookup of AACD.  If radix lookup returns just AA we can take the first three keys. and do the next lookup on ABFF.

File: sdk/helper/storagepacker/types.proto (hunk 18 above)
Comment (mgritter) [reply sdk/helper/storagepacker/types.proto]: Why was it present in your version? Is there an upgrade use case that I need to know about?

I *think* the upgrade will just be reading using SPv1 code and writing using SPv2 code, but the RFC didn't address it and there seem to be bits and pieces here where you seemed to be hinting at something else.

File: sdk/helper/storagepacker/types.proto (hunk 19 above)
Comment (mgritter) [reply sdk/helper/storagepacker/types.proto]: Item is still the return type, but it doesn't need to be a Protobuf type anymore. I can change that.

File: sdk/helper/storagepacker/partition.go (hunk 16 above)
Comment (mgritter) [reply sdk/helper/storagepacker/partition.go]: Oh, I forgot I did mention that alternative in the comment.

File: sdk/helper/storagepacker/ids2keys.go (hunk 20 above)
Comment (mgritter) [reply sdk/helper/storagepacker/ids2keys.go]: GetItems() takes a list of IDs. I want the returned items to respect the original order. So, this struct is a wrapper that maintains the initial request ordering for the items, as well as caching the key for each item ID.

Perhaps we could just use Item, but make `key` a non-exported member of the struct.

Or, it could be named something more verbose like wrappedRequest or wrappedItem.

File: sdk/helper/storagepacker/ids2keys.go (hunk 1 above)
Comment (mgritter) [reply sdk/helper/storagepacker/ids2keys.go]: I changed this to eliminate itemRequest, and just use Item everywhere (which now has a non-exported "key".)

File: sdk/helper/storagepacker/ids2keys.go (hunk 17 above)
Comment (mgritter) [reply sdk/helper/storagepacker/ids2keys.go]: Renamed these methods, split the file into keys.go and items.go.

File: sdk/helper/storagepacker/types.proto (hunk 18 above)
Comment (mgritter) [reply sdk/helper/storagepacker/types.proto]: Removed all the unnecessary fields, and changed Item to a normal type instead of a protobuf type.

File: sdk/helper/storagepacker/ids2keys.go (hunk 20 above)
Comment (mgritter) [reply sdk/helper/storagepacker/ids2keys.go]: Merged itemRequest and Item so we only use Item any more.

**Comment by vishalnayak** [comment]
Closing due to staleness. Reopen if required.
